File: tests/net/java/android/net/dnsproxy/DnsProxyServerTest.java
/*
 * Copyright (C) 2019 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package android.net.dnsproxy;

import static android.system.OsConstants.IPPROTO_TCP;
import static android.system.OsConstants.IPPROTO_UDP;

import static junit.framework.Assert.assertTrue;

import static org.mockito.AdditionalMatchers.aryEq;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.doNothing;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import android.annotation.NonNull;
import android.annotation.Nullable;
import android.net.Network;
import android.net.dnsproxy.DnsProxyServer.Dependencies;
import android.net.dnsproxy.DnsProxyServer.TCPListener;
import android.net.dnsproxy.DnsProxyServer.UDPListener;
import android.net.util.SharedLog;
import android.os.test.TestLooper;
import android.support.test.filters.SmallTest;
import android.support.test.runner.AndroidJUnit4;
import android.system.ErrnoException;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileDescriptor;
import java.io.FileOutputStream;
import java.net.InetAddress;
import java.util.Arrays;

@RunWith(AndroidJUnit4.class)
@SmallTest
public class DnsProxyServerTest {
    private static final String PROP_DEXMAKER_SHARE_CLASSLOADER = "dexmaker.share_classloader";
    private static final String TEST_IFACE = "testiface";
    private static final InetAddress TEST_CLIENT_ADDR = InetAddress.parseNumericAddress(
            "192.168.0.42");
    private static final int TEST_CLIENT_PORT = 572;
    private static final int TEST_NETID = 337;
    private static final int RCODE_FormErr = 1;
    private static final int RCODE_ServFail = 2;

    @NonNull
    @Mock
    private Dependencies mDeps;
    @NonNull
    @Mock
    private TCPListener mTCPListener;
    @NonNull
    @Mock
    private UDPListener mUDPListener;
    @NonNull
    private FileDescriptor mTestSocket;

    @NonNull
    @Captor
    private ArgumentCaptor<byte[]> mAnswerBuffCaptor;
    @NonNull
    @Captor
    private ArgumentCaptor<byte[]> mQueryBuffCaptor;
    @NonNull
    @Captor
    private ArgumentCaptor<InetAddress> mDstAddrCaptor;
    @NonNull
    @Captor
    private ArgumentCaptor<Integer> mDstPortCaptor;
    @NonNull
    @Captor
    private ArgumentCaptor<Integer> mProtoTypeCaptor;
    @NonNull
    @Captor
    private ArgumentCaptor<FileDescriptor> mFDCaptor;


    @NonNull
    private TestLooper mLooper;
    @NonNull
    private DnsProxyServer mServer;

    @Nullable
    private String mPrevShareClassloaderProp;

    @Before
    public void setUp() throws Exception {
        // Allow mocking package-private classes
        mPrevShareClassloaderProp = System.getProperty(PROP_DEXMAKER_SHARE_CLASSLOADER);
        System.setProperty(PROP_DEXMAKER_SHARE_CLASSLOADER, "true");

        MockitoAnnotations.initMocks(this);

        final FileOutputStream os = new FileOutputStream(new File("/dev/null"));
        mTestSocket = os.getFD();

        when(mDeps.makeUDPListener()).thenReturn(mUDPListener);
        when(mDeps.makeTCPListener()).thenReturn(mTCPListener);
        doNothing().when(mDeps).doQuery(mFDCaptor.capture(), mProtoTypeCaptor.capture(),
                mQueryBuffCaptor.capture(), mDstAddrCaptor.capture(), mDstPortCaptor.capture());
        doNothing().when(mDeps)
                .sendPacket(mFDCaptor.capture(), mProtoTypeCaptor.capture(),
                        mAnswerBuffCaptor.capture(), mDstAddrCaptor.capture(),
                        mDstPortCaptor.capture());

        when(mDeps.makeTCPListener()).thenReturn(mTCPListener);
        when(mDeps.makeUDPListener()).thenReturn(mUDPListener);

        mLooper = new TestLooper();
        mServer = new DnsProxyServer(mLooper.getLooper(), TEST_IFACE,
                new SharedLog(DnsProxyServerTest.class.getSimpleName()), mDeps);

        mServer.startWithNetwork(new Network(TEST_NETID));
        mLooper.dispatchAll();
    }

    @After
    public void tearDown() throws Exception {
        mServer.stop();
        mLooper.dispatchAll();
        System.setProperty(PROP_DEXMAKER_SHARE_CLASSLOADER,
                (mPrevShareClassloaderProp == null ? "" : mPrevShareClassloaderProp));
    }

    @Test
    public void testStart() throws Exception {
        verify(mTCPListener, times(1)).start();
        verify(mUDPListener, times(1)).start();
    }

    @Test
    public void testStop() throws Exception {
        mServer.stop();
        mLooper.dispatchAll();
        verify(mTCPListener, times(1)).stop();
        verify(mUDPListener, times(1)).stop();
    }

    @Test
    public void testProcessPacketNormaolQuery() throws Exception {
        // UDP
        byte[] packetBuf = getQueryPacket();

        mServer.processPacket(mTestSocket, IPPROTO_UDP, packetBuf, packetBuf.length,
                TEST_CLIENT_ADDR, TEST_CLIENT_PORT);
        verify(mDeps, times(1))
                .doQuery(eq(mTestSocket), eq(IPPROTO_UDP), aryEq(packetBuf),
                        eq(TEST_CLIENT_ADDR), eq(TEST_CLIENT_PORT));

        byte[] answerBuf = getAnswerPacket();

        mServer.transmitAnswer(mTestSocket, IPPROTO_UDP, answerBuf,
                TEST_CLIENT_ADDR, TEST_CLIENT_PORT);
        verify(mDeps, times(1))
                .sendPacket(eq(mTestSocket), eq(IPPROTO_UDP), aryEq(answerBuf),
                        eq(TEST_CLIENT_ADDR), eq(TEST_CLIENT_PORT));

        // TCP
        packetBuf = getQueryPacket();

        mServer.processPacket(mTestSocket, IPPROTO_TCP, packetBuf, packetBuf.length,
                TEST_CLIENT_ADDR, TEST_CLIENT_PORT);
        verify(mDeps, times(1))
                .doQuery(eq(mTestSocket), eq(IPPROTO_TCP), aryEq(packetBuf),
                        eq(TEST_CLIENT_ADDR), eq(TEST_CLIENT_PORT));

        answerBuf = getAnswerPacket();

        mServer.transmitAnswer(mTestSocket, IPPROTO_TCP, answerBuf,
                TEST_CLIENT_ADDR, TEST_CLIENT_PORT);
        verify(mDeps, times(1))
                .sendPacket(eq(mTestSocket), eq(IPPROTO_TCP), aryEq(addPacketLength(answerBuf)),
                        eq(TEST_CLIENT_ADDR), eq(TEST_CLIENT_PORT));
    }

    @Test
    public void testProcessPacketInvalidQuery() throws Exception {
        // UDP
        byte[] packetBuf = getInvalidPacket();
        mServer.processPacket(mTestSocket, IPPROTO_UDP, packetBuf, packetBuf.length,
                TEST_CLIENT_ADDR, TEST_CLIENT_PORT);

        verify(mDeps, never())
                .doQuery(any(), anyInt(), any(byte[].class), any(), anyInt());
        verify(mDeps, times(1))
                .sendPacket(eq(mTestSocket), eq(IPPROTO_UDP),
                        aryEq(setupFailResponse(packetBuf, RCODE_FormErr)),
                        eq(TEST_CLIENT_ADDR), eq(TEST_CLIENT_PORT));

        // TCP
        packetBuf = getInvalidPacket();
        mServer.processPacket(mTestSocket, IPPROTO_TCP, packetBuf, packetBuf.length,
                TEST_CLIENT_ADDR, TEST_CLIENT_PORT);

        verify(mDeps, never())
                .doQuery(any(), anyInt(), any(byte[].class), any(), anyInt());
        verify(mDeps, times(1))
                .sendPacket(eq(mTestSocket), eq(IPPROTO_TCP),
                        aryEq(addPacketLength(setupFailResponse(packetBuf, RCODE_FormErr))),
                        eq(TEST_CLIENT_ADDR), eq(TEST_CLIENT_PORT));
    }

    @Test
    public void testProcessMalformedPacket() throws Exception {
        // UDP
        byte[] packetBuf = getMalformedPacket();

        mServer.processPacket(mTestSocket, IPPROTO_UDP, packetBuf, packetBuf.length,
                TEST_CLIENT_ADDR, TEST_CLIENT_PORT);

        verify(mDeps, never()).doQuery(
                any(), anyInt(), any(byte[].class), any(), anyInt());
        verify(mDeps, never()).sendPacket(
                any(), anyInt(), any(byte[].class), any(), anyInt());

        // TCP
        mServer.processPacket(mTestSocket, IPPROTO_TCP, packetBuf, packetBuf.length,
                TEST_CLIENT_ADDR, TEST_CLIENT_PORT);

        verify(mDeps, never()).doQuery(
                any(), anyInt(), any(byte[].class), any(), anyInt());
        verify(mDeps, never()).sendPacket(
                any(), anyInt(), any(byte[].class), any(), anyInt());
    }

    @Test
    public void testProcessPacketHandleQueryFail() throws Exception {
        doAnswer(inv -> {
            throw new ErrnoException("Test Errno", 100 /* ENETDOWN */);
        }).when(mDeps).doQuery(any(), anyInt(), any(byte[].class), any(), anyInt());

        // UDP
        byte[] packetBuf = getQueryPacket();
        mServer.processPacket(mTestSocket, IPPROTO_UDP, packetBuf, packetBuf.length,
                TEST_CLIENT_ADDR, TEST_CLIENT_PORT);

        verify(mDeps, times(1))
                .doQuery(any(), anyInt(), any(byte[].class), any(), anyInt());
        verify(mDeps, times(1))
                .sendPacket(eq(mTestSocket), eq(IPPROTO_UDP),
                        aryEq(setupFailResponse(packetBuf, RCODE_ServFail)),
                        eq(TEST_CLIENT_ADDR), eq(TEST_CLIENT_PORT));

        // TCP
        packetBuf = getQueryPacket();
        mServer.processPacket(mTestSocket, IPPROTO_TCP, packetBuf, packetBuf.length,
                TEST_CLIENT_ADDR, TEST_CLIENT_PORT);

        verify(mDeps, times(2))
                .doQuery(any(), anyInt(), any(byte[].class), any(), anyInt());
        verify(mDeps, times(1))
                .sendPacket(eq(mTestSocket), eq(IPPROTO_TCP),
                        aryEq(addPacketLength(setupFailResponse(packetBuf, RCODE_ServFail))),
                        eq(TEST_CLIENT_ADDR), eq(TEST_CLIENT_PORT));
    }

    @Test
    public void testTransmitAnswerNormal() throws Exception {
        // UDP
        byte[] packetBuf = getAnswerPacket();
        mServer.transmitAnswer(mTestSocket, IPPROTO_UDP, packetBuf,
                TEST_CLIENT_ADDR, TEST_CLIENT_PORT);

        verify(mDeps, never())
                .doQuery(any(), anyInt(), any(byte[].class), any(), anyInt());
        verify(mDeps, times(1))
                .sendPacket(eq(mTestSocket), eq(IPPROTO_UDP), aryEq(packetBuf),
                        eq(TEST_CLIENT_ADDR), eq(TEST_CLIENT_PORT));

        // TCP
        packetBuf = getAnswerPacket();
        mServer.transmitAnswer(mTestSocket, IPPROTO_TCP, packetBuf,
                TEST_CLIENT_ADDR, TEST_CLIENT_PORT);

        verify(mDeps, never())
                .doQuery(any(), anyInt(), any(byte[].class), any(), anyInt());
        verify(mDeps, times(1))
                .sendPacket(eq(mTestSocket), eq(IPPROTO_TCP),
                        aryEq(addPacketLength(packetBuf)), eq(TEST_CLIENT_ADDR),
                        eq(TEST_CLIENT_PORT));
    }

    @Test
    public void testTransmitAnswerHandleSendFail() throws Exception {
        doAnswer(inv -> {
            throw new ErrnoException("Test Errno", 100 /* ENETDOWN */);
        }).when(mDeps).sendPacket(any(), anyInt(), any(byte[].class), any(), anyInt());

        // UDP
        byte[] packetBuf = getAnswerPacket();
        mServer.transmitAnswer(mTestSocket, IPPROTO_UDP, packetBuf,
                TEST_CLIENT_ADDR, TEST_CLIENT_PORT);

        verify(mDeps, never())
                .doQuery(any(), anyInt(), any(byte[].class), any(), anyInt());
        verify(mDeps, times(1))
                .sendPacket(eq(mTestSocket), eq(IPPROTO_UDP),
                        aryEq(setupFailResponse(packetBuf, RCODE_ServFail)),
                        eq(TEST_CLIENT_ADDR), eq(TEST_CLIENT_PORT));

        // TCP
        packetBuf = getAnswerPacket();
        mServer.transmitAnswer(mTestSocket, IPPROTO_TCP, packetBuf,
                TEST_CLIENT_ADDR, TEST_CLIENT_PORT);

        verify(mDeps, never())
                .doQuery(any(), anyInt(), any(byte[].class), any(), anyInt());
        verify(mDeps, times(1))
                .sendPacket(eq(mTestSocket), eq(IPPROTO_TCP),
                        aryEq(addPacketLength(packetBuf)),
                        eq(TEST_CLIENT_ADDR), eq(TEST_CLIENT_PORT));

    }

    // TODO: Extend DnsPacket and use it to create test packets.
    private byte[] getQueryPacket() {
        return new byte[]{
                /* Header */
                0x55, 0x66, /* Transaction ID */
                0x10, 0x00, /* Flags */
                0x00, 0x01, /* Questions */
                0x00, 0x00, /* Answer RRs */
                0x00, 0x00, /* Authority RRs */
                0x00, 0x00, /* Additional RRs */
                /* Queries */
                0x03, 0x77, 0x77, 0x77, 0x06, 0x67, 0x6F, 0x6F, 0x67, 0x6c, 0x65,
                0x03, 0x63, 0x6f, 0x6d, 0x00, /* Name */
                0x00, 0x01, /* Type */
                0x00, 0x01, /* Class */
        };
    }

    private byte[] getMalformedPacket() {
        // Packet size < DNS header
        return new byte[]{
                0x35, 0x36,
                0x21, 0x50,
                0x20, 0x01,
                0x11, 0x05,
        };
    }

    private byte[] getInvalidPacket() {
        return new byte[]{
                /* Header */
                0x55, 0x66, /* Transaction ID */
                (byte) 0x81, (byte) 0x80, /* Flags */
                0x00, 0x01, /* Questions */
                0x00, 0x00, /* Answer RRs */
                0x00, 0x00, /* Authority RRs */
                0x00, 0x00, /* Additional RRs */
                /* Queries */
                0x03, 0x77, 0x77, 0x77, 0x06, 0x67, 0x6F, 0x6F, 0x67, 0x6c, 0x65,
                0x03, 0x63, 0x6f, 0x6d, 0x00, /* Name */
                0x00, 0x01, /* Type */
                0x00, 0x01, /* Class */
        };
    }

    private byte[] getAnswerPacket() {
        return new byte[]{
                /* Header */
                0x55, 0x66, /* Transaction ID */
                (byte) 0x81, (byte) 0x80, /* Flags */
                0x00, 0x01, /* Questions */
                0x00, 0x01, /* Answer RRs */
                0x00, 0x00, /* Authority RRs */
                0x00, 0x00, /* Additional RRs */
                /* Queries */
                0x03, 0x77, 0x77, 0x77, 0x06, 0x67, 0x6F, 0x6F, 0x67, 0x6c, 0x65,
                0x03, 0x63, 0x6f, 0x6d, 0x00, /* Name */
                0x00, 0x01, /* Type */
                0x00, 0x01, /* Class */
                /* Answers */
                (byte) 0xc0, 0x0c, /* Name */
                0x00, 0x01, /* Type */
                0x00, 0x01, /* Class */
                0x00, 0x00, 0x01, 0x2b, /* TTL */
                0x00, 0x04, /* Data length */
                (byte) 0xac, (byte) 0xd9, (byte) 0xa1, (byte) 0x84 /* Address */
        };
    }

    private static void assertArrayEquals(byte[] expected, byte[] actual) {
        assertTrue("Expected=" + Arrays.toString(expected) + ", actual=" + Arrays.toString(actual),
                Arrays.equals(expected, actual));
    }

    private byte[] addPacketLength(@NonNull byte[] buf) throws Exception {
        final byte[] ansPacket = new byte[buf.length + 2];
        ansPacket[0] = (byte) ((buf.length >> 8) & 0xff);
        ansPacket[1] = (byte) (buf.length & 0xff);
        System.arraycopy(buf, 0, ansPacket, 2, buf.length);
        return ansPacket;
    }

    private byte[] setupFailResponse(@NonNull byte[] querybuf, int rcode) {
        // Change rcode and set OR.
        // TODO: Refine DnsProxyPacket to create answer.
        querybuf[2] = (byte) (querybuf[2] | 0x80);
        querybuf[3] = (byte) rcode;
        return querybuf;
    }
}

File: packages/Tethering/tests/unit/src/android/net/dnsproxy/DnsProxyServerTest.kt
/*
 * Copyright (C) 2020 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package android.net.dnsproxy

import android.net.DnsResolver.FLAG_NO_RETRY
import android.net.dnsproxy.DnsProxyServer.addPacketLength
import android.net.dnsproxy.DnsProxyServer.setupFailResponse
import android.system.OsConstants.EBADF

import junit.framework.Assert.assertFalse
import junit.framework.Assert.assertTrue

import org.mockito.AdditionalMatchers.aryEq
import org.mockito.ArgumentMatchers.any
import org.mockito.ArgumentMatchers.anyInt
import org.mockito.ArgumentMatchers.eq
import org.mockito.Mockito
import org.mockito.Mockito.doAnswer
import org.mockito.Mockito.doReturn
import org.mockito.Mockito.mock
import org.mockito.Mockito.never
import org.mockito.Mockito.reset
import org.mockito.Mockito.spy
import org.mockito.Mockito.times
import org.mockito.Mockito.verify
import android.net.DnsResolver
import android.net.Network
import android.net.dnsproxy.DnsProxyServer.ResponderDependencies
import android.net.dnsproxy.DnsProxyServer.ServerDependencies
import android.net.util.SharedLog
import android.os.HandlerThread
import android.os.MessageQueue
import android.system.ErrnoException
import android.system.Os
import android.system.OsConstants
import android.testing.AndroidTestingRunner
import android.testing.TestableLooper
import android.testing.TestableLooper.RunWithLooper

import androidx.test.filters.SmallTest

import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith

import java.io.FileDescriptor
import java.io.IOException
import java.net.InetAddress
import java.net.InetSocketAddress
import java.util.Arrays
import java.util.concurrent.Executor

private val TEST_IFACE = "testIface"
private val TEST_CLIENT_ADDR = InetAddress.parseNumericAddress("192.168.0.42")
private val TEST_CLIENT_PORT = 572
private val RCODE_FORMERR = 1
private val QUERY_FLAGS = FLAG_NO_RETRY

@RunWith(AndroidTestingRunner::class)
@SmallTest
@RunWithLooper
class DnsProxyServerTest {
    private val mockResponderDeps = mock(ResponderDependencies::class.java)
    private val mockQueue = mock(MessageQueue::class.java)
    private val mockNetwork = mock(Network::class.java)
    private val mockDnsResolver = mock(DnsResolver::class.java)
    private val mockLog = mock(SharedLog::class.java)
    private var mockInetSocketAddress = mock(InetSocketAddress::class.java)
    private val mockTestExecutor = mock(Executor::class.java)
    private var mockHandlerThread = mock(HandlerThread::class.java)

    private var mTestTCPSocket = spy(Os.open("/dev/null", 0, OsConstants.O_RDONLY))
    private var mTestUDPSocket = spy(Os.open("/dev/null", 0, OsConstants.O_RDONLY))
    private lateinit var mLooper: TestableLooper
    private lateinit var mServer: DnsProxyServer

    private val mockServerDeps = mock(ServerDependencies::class.java).apply {
        doReturn(mTestUDPSocket).`when`(this).makeUDPSocket()
        doReturn(mTestTCPSocket).`when`(this).makeTCPSocket()
        doReturn(mockHandlerThread).`when`(this).makeHandlerThread(any())
    }

    // TODO: Extend DnsPacket and use it to create test packets.
    private val queryPacket = byteArrayOf(
            /* Header */
            0x55, 0x66, /* Transaction ID */
            0x10, 0x00, /* Flags */
            0x00, 0x01, /* Questions */
            0x00, 0x00, /* Answer RRs */
            0x00, 0x00, /* Authority RRs */
            0x00, 0x00, /* Additional RRs */
            /* Queries */
            0x03, 0x77, 0x77, 0x77, 0x06, 0x67, 0x6F, 0x6F, 0x67, 0x6c, 0x65,
            0x03, 0x63, 0x6f, 0x6d, 0x00, /* Name */
            0x00, 0x01, /* Type */
            0x00, 0x01) /* Class */

    private val invalidPacket = byteArrayOf(
            /* Header */
            0x55, 0x66, /* Transaction ID */
            0x81.toByte(), 0x80.toByte(), /* Flags */
            0x00, 0x01, /* Questions */
            0x00, 0x00, /* Answer RRs */
            0x00, 0x00, /* Authority RRs */
            0x00, 0x00, /* Additional RRs */
            /* Queries */
            0x03, 0x77, 0x77, 0x77, 0x06, 0x67, 0x6F, 0x6F, 0x67, 0x6c, 0x65,
            0x03, 0x63, 0x6f, 0x6d, 0x00, /* Name */
            0x00, 0x01, /* Type */
            0x00, 0x01) /* Class */

    private val answerPacket = byteArrayOf(
            /* Header */
            0x55, 0x66, /* Transaction ID */
            0x81.toByte(), 0x80.toByte(), /* Flags */
            0x00, 0x01, /* Questions */
            0x00, 0x01, /* Answer RRs */
            0x00, 0x00, /* Authority RRs */
            0x00, 0x00, /* Additional RRs */
            /* Queries */
            0x03, 0x77, 0x77, 0x77, 0x06, 0x67, 0x6F, 0x6F, 0x67, 0x6c, 0x65,
            0x03, 0x63, 0x6f, 0x6d, 0x00, /* Name */
            0x00, 0x01, /* Type */
            0x00, 0x01, /* Class */
            /* Answers */
            0xc0.toByte(), 0x0c, /* Name */
            0x00, 0x01, /* Type */
            0x00, 0x01, /* Class */
            0x00, 0x00, 0x01, 0x2b, /* TTL */
            0x00, 0x04, /* Data length */
            0xac.toByte(), 0xd9.toByte(), 0xa1.toByte(), 0x84.toByte()) /* Address */

    // Packet size < DNS header
    private val malformedPacket = byteArrayOf(0x35, 0x36, 0x21, 0x50, 0x20, 0x01, 0x11, 0x05)

    @Before
    @Throws(Exception::class)
    fun setUp() {
        mLooper = TestableLooper.get(this)
        val looper = spy(mLooper.looper)
        doReturn(mockQueue).`when`(looper).queue
        doReturn(looper).`when`(mockHandlerThread).looper

        mServer = DnsProxyServer(TEST_IFACE,
                SharedLog(DnsProxyServerTest::class.java.simpleName), mockServerDeps)
        mServer.start(mockNetwork)
        mLooper.processAllMessages()
    }

    @After
    @Throws(Exception::class)
    fun tearDown() {
        mServer.stop()
        mLooper.processMessages(1)
        verify(mockQueue, times(1)).removeOnFileDescriptorEventListener(eq(mTestUDPSocket))
        verify(mockQueue, times(1)).removeOnFileDescriptorEventListener(eq(mTestTCPSocket))
        assertTrue(!mTestUDPSocket.valid())
        assertTrue(!mTestTCPSocket.valid())
        verify(mockHandlerThread, times(1)).quitSafely()
    }

    @Test
    fun testStart() {
        verify(mockServerDeps, times(1)).listenDnsUDP(eq(mTestUDPSocket), any())
        verify(mockServerDeps, times(1)).listenDnsTCP(eq(mTestTCPSocket), any())
        verify(mockQueue, times(1))
                .addOnFileDescriptorEventListener(eq(mTestUDPSocket), anyInt(), any())
        verify(mockQueue, times(1))
                .addOnFileDescriptorEventListener(eq(mTestTCPSocket), anyInt(), any())
    }

    @Test
    fun testUpdateUpstream() {
        assertTrue(mServer.isUpstreamAvailable)
        mServer.updateUpstream(null)
        mLooper.processMessages(1)
        assertFalse(mServer.isUpstreamAvailable)
        mServer.updateUpstream(mockNetwork)
        mLooper.processMessages(1)
        assertTrue(mServer.isUpstreamAvailable)
    }

    @Test
    fun testRequestResponder() {
        val request = spy(object : DnsProxyServer.RequestResponder(
                mockNetwork, mTestUDPSocket, mockLog,
                mockResponderDeps, mockDnsResolver, mockTestExecutor) {
            internal override fun transmitAnswer(
                fd: FileDescriptor?,
                buf: ByteArray?,
                dstAddr: InetAddress?,
                dstPort: Int
            ) {
            }

            @Throws(IOException::class, ErrnoException::class)
            internal override fun readPacket(fd: FileDescriptor?): ByteArray {
                mDstAddr = mockInetSocketAddress
                return queryPacket
            }
                })

        // handleRequest, exception case
        Mockito.`when`(request.readPacket(any())).thenThrow(ErrnoException("Test Errno", EBADF))
        request.handleRequest()
        verify(request, never()).processDnsPacket(any(), any(), any(), any(), anyInt())
        verify(request, times(1)).cleanup(eq(mTestUDPSocket))

        // handleRequest, normal case
        reset(request)
        request.handleRequest()
        verify(request, never()).processDnsPacket(eq(mockNetwork), eq(mTestUDPSocket),
                aryEq(queryPacket), eq(TEST_CLIENT_ADDR), eq(TEST_CLIENT_PORT))
        verify(mockDnsResolver, times(1))
                .rawQuery(eq(mockNetwork), aryEq(queryPacket), eq(QUERY_FLAGS),
                        eq(mockTestExecutor), any(), any())

        // Process NormalQuery
        reset(request)
        reset(mockDnsResolver)
        var packetBuf = queryPacket
        request.processDnsPacket(mockNetwork, mTestUDPSocket, packetBuf,
                TEST_CLIENT_ADDR, TEST_CLIENT_PORT)
        verify(mockDnsResolver, times(1))
                .rawQuery(eq(mockNetwork), aryEq(packetBuf), eq(QUERY_FLAGS), eq(mockTestExecutor),
                        any(), any())
        reset(mockDnsResolver)
        // Process InvalidQuery
        packetBuf = invalidPacket
        request.processDnsPacket(mockNetwork, mTestUDPSocket, packetBuf,
                TEST_CLIENT_ADDR, TEST_CLIENT_PORT)
        verify(mockDnsResolver, never()).rawQuery(any(), any(), anyInt(), any(), any(), any())
        verify(request, times(1))
                .transmitAnswer(eq(mTestUDPSocket),
                        aryEq(setupFailResponse(packetBuf, RCODE_FORMERR)),
                        eq(TEST_CLIENT_ADDR), eq(TEST_CLIENT_PORT))

        // Process MalformedPacket
        packetBuf = malformedPacket
        request.processDnsPacket(mockNetwork, mTestUDPSocket, packetBuf,
                TEST_CLIENT_ADDR, TEST_CLIENT_PORT)
        verify(mockDnsResolver, never()).rawQuery(any(), any(), anyInt(), any(), any(), any())
        verify(request, times(1)).cleanup(eq(mTestUDPSocket))
    }

    @Test
    fun testUDPRequestResponder() {
        val answerBuf = answerPacket
        doAnswer { inv ->
            val buf = inv.getArgument<ByteArray>(1)
            System.arraycopy(buf, 0, answerBuf, 0, answerBuf.size)
            answerBuf.size
        }.`when`(mockResponderDeps).recvfrom(any(), any(),
                eq(DnsProxyServer.UDPRequestResponder.UDP_BUFFER_LENGTH), any())
        val request = DnsProxyServer.UDPRequestResponder(
                mockNetwork, mTestUDPSocket, mockLog,
                mockResponderDeps, mockDnsResolver, mockTestExecutor)
        assertTrue(request is DnsProxyServer.RequestResponder)

        request.transmitAnswer(mTestUDPSocket, answerBuf, TEST_CLIENT_ADDR, TEST_CLIENT_PORT)
        verify(mockResponderDeps, times(1))
                .sendPacket(eq(mTestUDPSocket), aryEq(answerBuf), eq(TEST_CLIENT_ADDR),
                        eq(TEST_CLIENT_PORT))

        val buf = request.readPacket(mTestUDPSocket)
        verify(mockResponderDeps, times(1))
                .recvfrom(eq(mTestUDPSocket), any(),
                        eq(DnsProxyServer.UDPRequestResponder.UDP_BUFFER_LENGTH),
                        any())
        assertTrue(Arrays.equals(buf, answerBuf))
    }

    @Test
    fun testTCPRequestResponder() {
        val answerBuf = answerPacket
        // Return length of packet
        doAnswer { inv ->
            val buf = inv.getArgument<ByteArray>(1)
            buf[0] = (answerBuf.size and 0xff00).toByte()
            buf[1] = (answerBuf.size and 0xff).toByte()
            2
        }.`when`(mockResponderDeps).read(any(), any(), eq(0), eq(2))
        // Return packet
        doAnswer { inv ->
            val buf = inv.getArgument<ByteArray>(1)
            System.arraycopy(buf, 0, answerBuf, 0, answerBuf.size)
            answerBuf.size
        }.`when`(mockResponderDeps).read(any(), any(), eq(0), eq(answerBuf.size))

        val request = spy(DnsProxyServer.TCPRequestResponder(
                mockNetwork, mTestTCPSocket,
                mockInetSocketAddress, mockLog, mockResponderDeps,
                mockDnsResolver, mockTestExecutor))
        assertTrue(request is DnsProxyServer.RequestResponder)

        request.transmitAnswer(mTestTCPSocket, answerBuf, TEST_CLIENT_ADDR, TEST_CLIENT_PORT)
        verify(mockResponderDeps, times(1))
                .sendPacket(eq(mTestTCPSocket), aryEq(addPacketLength(answerBuf)),
                        eq(TEST_CLIENT_ADDR),
                        eq(TEST_CLIENT_PORT))
        verify(request, times(1)).cleanup(eq(mTestTCPSocket))

        val buf = request.readPacket(mTestTCPSocket)
        verify(request, times(1)).setSocketTimeout(eq(mTestTCPSocket))
        verify(mockResponderDeps, times(1)).read(eq(mTestTCPSocket), any(), eq(0), eq(2))
        verify(mockResponderDeps, times(1))
                .read(eq(mTestTCPSocket), any(), eq(0), eq(answerBuf.size))
        assertTrue(Arrays.equals(buf, answerBuf))
    }
}
